src/components/implementation/tests/micro_booter/mb_tests.c

```diff
@@ -431,6 +442,53 @@ test_budgets(void)
 	PRINTC("Done.\n");
 }
 
+static void
+test_budgets_multi(void)
+{
+	int i;
+
+	PRINTC("Starting multi-level budget test.\n");
+
+	exec_cluster_alloc(&mbt.p, spinner_cyc,  &(mbt.p.cyc), BOOT_CAPTBL_SELF_INITRCV_BASE);
+	exec_cluster_alloc(&mbt.c, spinner_cyc, &(mbt.c.cyc), mbt.p.rc);
+	exec_cluster_alloc(&mbt.g, spinner_cyc, &(mbt.g.cyc), mbt.c.rc);
+
+	PRINTC("Budget switch latencies:\n");
+	for (i = 1 ; i < 10 ; i++) {
+		tcap_res_t res;
+		thdid_t  tid;
+		int      rcving;
+		cycles_t cycles, s, e;
+
+		/* test both increasing budgets and constant budgets */
+		if (i > 5) res = 1600000;
```

Comment:
not a fan of the inline constants.  If you give them a `#define` name, then people know what they are.